core/core.go

```diff
@@ -731,8 +739,35 @@ func constructPeerHost(ctx context.Context, id peer.ID, ps pstore.Peerstore, bwr
 		hostOpts = append(hostOpts, p2pbhost.NATPortMap)
 	}
 
+	if !opts.DisableRelay {
+		filterRelayAddr := func(addrs []ma.Multiaddr) []ma.Multiaddr {
+			var raddrs []ma.Multiaddr
+			for _, addr := range addrs {
+				_, err := addr.ValueForProtocol(circuit.P_CIRCUIT)
+				if err == nil {
+					continue
+				}
+				raddrs = append(raddrs, addr)
+			}
+			return raddrs
+		}
+		hostOpts = append(hostOpts, p2pbhost.AddrsFactory(filterRelayAddr))
```

Comment:
I will need to add the `AddrsFactory` option to `ConstructPeerHostOpts` in order to scaffold the composition properly.

Then we have two options for the composition: filter first or announce factory first. It makes some sense to use the filter first, the announce factory might actually want to announce some relay addresses in the future. On the other hand it also makes sense for the announce factory to see the p2p-circuit relay address, so that it knows that relay is activated.